### PR TITLE
Load all symbols from HDF5

### PR DESCRIPTION
**Pull request by james-s-willis**
… these symbols: H5E_CANTREGISTER_g, H5E_CALLBACK_g, H5E_PLINE_g, H5E_ERR_CLS_g when h5py is installed with --no-binary=h5py.

Addresses: https://github.com/kiyo-masui/bitshuffle/issues/105

### CODE REVIEW

File: src/hdf5_dl.c [hunk 1]
```diff
@@ -97,44 +97,167 @@ hid_t H5E_ERR_CLS_g = -1;
...
     int retval = -1;
-  	void * handle;
+  	void *handle = NULL, *symbol = NULL;
 
     handle = dlopen(libname, RTLD_LAZY | RTLD_LOCAL);
 
     if (handle != NULL) {
         /*H5*/
-        DL_H5Functions.H5open = (DL_func_H5open)dlsym(handle, "H5open");
+        if(DL_H5Functions.H5open == NULL) {
```
Comment (jrs65) [inline src/hdf5_dl.c:134]: These repeated blocks were a little hard to follow. It seems there are two things being done at the same time which prevent things being more neatly structured:
- Attempt to resolve a symbol if it is not already resolved
- Check if all symbols are now resolved

I can see it being a lot cleaner to completely separate these, i.e. you have one pass doing the first, and then after you've done that an entirely separate pass (probably placed in a separate function) which checks that all the symbols have been resolved.

File: src/hdf5_dl.c [hunk 2]
```diff
@@ -97,44 +97,167 @@ hid_t H5E_ERR_CLS_g = -1;
...
  *
  * libname: Name of the DLL from which to load libHDF5 symbols
  * Returns: a value < 0 if an error occured
  */
-int init_filter(const char* libname)
+int init_filter(const char *libname)
 {
     int retval = -1;
-  	void * handle;
+  	void *handle = NULL, *symbol = NULL;
```
Comment (jrs65) [inline src/hdf5_dl.c:128]: The meaning of the return value in this function is inconsistent with the documentation, although it looks like this was the same in the previous implementation, i.e. you return -1 if all symbols have been successfully resolved, which I would consider a success, but according to the docstring (and C conventions) should indicate a failure.

**Review by jrs65 — APPROVED** [review]
This looks great. Thanks James. I added one comment about the return value which I think should be fixed somehow, and another optional one on the structuring, as it's become a little hard to follow the logic as it's grown.

**Review by anjakefala — APPROVED** [review]
I really appreciate you looking into and addressing this!! ^^

File: bitshuffle/h5.pyx [hunk 3]
```diff
@@ -65,10 +65,24 @@ H5_COMPRESS_LZ4 = BSHUF_H5_COMPRESS_LZ4
 # Init HDF5 dynamic loading with HDF5 library used by h5py
 if not sys.platform.startswith('win'):
     if sys.version_info[0] >= 3:
-        libname = bytes(h5d.__file__, encoding='utf-8')
+        libs = [bytes(h5d.__file__, encoding='utf-8'),
+                bytes(h5s.__file__, encoding='utf-8'),
+                bytes(h5t.__file__, encoding='utf-8'),
+                bytes(h5p.__file__, encoding='utf-8'),
+                bytes(h5z.__file__, encoding='utf-8')]
```
Comment (ketiltrout) [inline bitshuffle/h5.pyx:72]: I'd also include `h5py.defs` and `h5py.h5fd`, which are the other two I found that were linked to libhdf5

File: src/hdf5_dl.c [hunk 4]
```diff
@@ -97,44 +97,167 @@ hid_t H5E_ERR_CLS_g = -1;
...
+        }
 
         /*Variables*/
-        H5E_CANTREGISTER_g = *((hid_t *)dlsym(handle, "H5E_CANTREGISTER_g"));
-        H5E_CALLBACK_g = *((hid_t *)dlsym(handle, "H5E_CALLBACK_g"));
-        H5E_PLINE_g = *((hid_t *)dlsym(handle, "H5E_PLINE_g"));
-        H5E_ERR_CLS_g = *((hid_t *)dlsym(handle, "H5E_ERR_CLS_g"));
-
-        is_init = true;
+        if(H5E_CANTREGISTER_g == -1) {
```
Comment (ketiltrout) [inline src/hdf5_dl.c:227]: You're assuming the HDF5 devs won't ever assign the value -1 to this symbol (or any of the other `..._g` variables) in their library, which I don't think you can do.  There's no value of this variable which you can use as a sigil to indicate it hasn't been resolved because there are no values available to you that aren't available to them.  You need something else out-of-band to track whether the value is properly initialised or not.

Comment (ketiltrout) [reply src/hdf5_dl.c:268]: The easiest might be to just collect all the pointers for these variables (the ones return by dlsym), which you _can_ distinguish from NULL, and then just dereference them all at the end once you've determined you've found everything.

Comment (james-s-willis) [reply src/hdf5_dl.c]: @ketiltrout, I've added pointers to address this issue now

File: src/hdf5_dl.c [hunk 5]
```diff
@@ -97,44 +105,209 @@ hid_t H5E_ERR_CLS_g = -1;
...
+}
+
+/*
+ * Check that all symbols have been loaded
+ * 
+ * Returns: 0 if an error occured, -1 for success
+ */
+int check_symbols() {
+
+  int retval = -1;
```
Comment (ketiltrout) [inline src/hdf5_dl.c:135]: You don't need `retval`.  Short circuit the tests: use `return 0` instead `retval = 0` and then a `return -1` at the end to cover the successful case.

File: src/hdf5_dl.c [hunk 6]
```diff
@@ -97,44 +105,209 @@ hid_t H5E_ERR_CLS_g = -1;
...
+  if(ret != NULL && (err = dlerror()) == NULL)
+    return ret;
+  else
+    return NULL;
+}
+
+/*
+ * Check that all symbols have been loaded
+ * 
+ * Returns: 0 if an error occured, -1 for success
```
Comment (ketiltrout) [inline src/hdf5_dl.c:131]: This is the opposite of the standard C convention (where zero indicates success).  I'd consider reversing the logic, but if you'd prefer to keep it this way, you should at least point out that it's reversed.

File: src/hdf5_dl.c [hunk 7]
```diff
@@ -97,44 +105,209 @@ hid_t H5E_ERR_CLS_g = -1;
...
+
+  return retval;
+
+}
 
 /* Initialize the dynamic loading of symbols and register the plugin
  *
  * libname: Name of the DLL from which to load libHDF5 symbols
- * Returns: a value < 0 if an error occured
+ * Returns: 0 if an error occured, -1 for success
```
Comment (ketiltrout) [inline src/hdf5_dl.c:194]: Here, too, people familiar with C will assume 0==success.

File: src/hdf5_dl.c [hunk 8]
```diff
@@ -97,44 +105,209 @@ hid_t H5E_ERR_CLS_g = -1;
...
+  
+  if(DL_H5Functions.H5Tclose == NULL)
+    retval = 0;
+
+  /*H5Z*/
+  if(DL_H5Functions.H5Zregister == NULL)
+    retval = 0;
+
+  /*Variables*/
+  if(H5E_CANTREGISTER_g == -1)
```
Comment (ketiltrout) [inline src/hdf5_dl.c:175]: This is the same problem as before: -1 is a value that might have been assigned to this value by libhdf5.  You need to check whether the pointer is NULL or not.

**Review by ketiltrout — CHANGES_REQUESTED** [review]
Your variable checks still seem wrong: you're still checking the value of the variable rather than the pointers to determine if they've been resolved.

Also I'd strongly recommend reversing the return values of the functions to align to the C convention of 0==success .  If you don't, you need to at least point out that the you're doing the opposite of the convention because people familiar with C are going to be confused.

**Comment by james-s-willis** [comment]
@ketiltrout I've addressed your last round of comments

File: src/hdf5_dl.c [hunk 9]
```diff
@@ -97,44 +105,204 @@ hid_t H5E_ERR_CLS_g = -1;
...
+        }
+
         /*H5E*/
-        DL_H5Functions.H5Epush1 = (DL_func_H5Epush1)dlsym(handle, "H5Epush1");
-        DL_H5Functions.H5Epush2 = (DL_func_H5Epush2)dlsym(handle, "H5Epush2");
+        if(DL_H5Functions.H5Epush1 == NULL) {
+            symbol = find_sym(handle, "H5Epush1");
+            if(symbol != NULL)
+                DL_H5Functions.H5Epush1 = (DL_func_H5Epush1)symbol;
+        }
```
Comment (jrs65) [inline src/hdf5_dl.c:214]: You could consider some ways to shorten this boiler plate:
- Wrap it up in a function or macro
- Don't bother with the conditional assignment as it doesn't do anything (i.e. you know the struct entry is NULL, so when the symbol is NULL, assigning it to the entry does nothing). Maybe leave a comment to that effect on the first one as not checking the value could look like a bug.

TBH if you do the latter, there's probably not enough boilerplate left to warrant the former.

**Review by jrs65 — CHANGES_REQUESTED** [review]

File: src/hdf5_dl.c (hunk 9 above)
Comment (james-s-willis) [reply src/hdf5_dl.c]: Fixed this now.

**Review by jrs65 — APPROVED** [review]
